src/sources/host_metrics.rs

```diff
@@ -692,6 +694,22 @@ fn add_collector(collector: &str, mut metrics: Vec<Metric>) -> Vec<Metric> {
     metrics
 }
 
+#[cfg(target_os = "linux")]
+pub fn init_roots() {
+    let virt = futures::executor::block_on(heim::virt::detect());
```

Comment:
Should this be an `async fn` to just use `.await` here?

Reply:
We _can_ make this async, but then any synchronous function that calls this one like we're doing in `app.rs` has an unused implementer of `futures::Future` which breaks our lints

Reply:
Right, we would have to move the blocking into the caller(s), which would require them to either become async themselves or move the `block_on` there.

Reply:
Ah indeed - I more meant we have to block on it somewhere, so if you'd prefer to do so at the call site just say the word.

Reply:
Hmm, looking at it further, I'm surprised this works. AFAIK `block_on` depends on an async runtime being running, but the runtime isn't created until later in `fn prepare_from_opts`. I would have expected this to throw an error.

In any case, this kind of issue is why I would tend to want to move the blocking into the caller, particularly for early init functions, to make it obvious that a runtime must be running.

Reply:
I think that's a very good call @bruceg ! I'll push a fix first thing in the am

Reply:
The reason it works is it’s not a `tokio` executor, but `futures` - it simply run the future to completion on the current thread (without `spawn` support, but with the ability to manage concurrent execution within one feature).
@eeyun I didn’t expect this to involve async operation, I assumed we’d just assign values from the env syncronously. I still think we don’t need to go beyond that.

Reply:
Not sure what to tell you about that. Nothing in the heim functions we added is explicitly async so I was surprised by this myself.

Reply:
`fn heim::virt::detect` is async, and digging a bit, what it calls involves some async filesystem calls (reads from proc and sys, etc). Since you removed that bit, there is no longer any need for this to be async, and so no `block_on`.

@MOZGIII I'm confused then. Without tokio, what handles things like non-blocking I/O that are invoked by the futures?

Reply:
Non-blocking io can still function as long as there's a reactor loop. (I suspect the terminology I'm used to is different from the naming conventions used in `tokio` - so take this with a grain of salt).
Reactor loop is not necessarily a scheduler and doesn't necessarily handle "task" management and lifecycle - it just has to poll the async function when the Waker signals, and park the thread when the future returns pending. Afaik, In the case of async io, the Waker abstraction is enough to handle the progress.

We essentially only need something like tokio if we want to schedule tasks over a thread pool with a scheduler, or if we need a timer (which doesn't have to be handled at the reactor level, but in the current ecosystem it is). IO operations are portable across the "async runtimes".